File: app/web/src/lib/running.ts
import type { Item, Person, RunningRow } from './types';
import { computeSharesForItem } from './splits';
import { round2 } from './math';

export function buildRunningRows(
  people: Person[],
  items: Item[],
  summary: {
    subtotals: number[];
    taxShares: number[];
    tipShares: number[];
    totals: number[];
  },
  taxAmount: number,
  tipAmount: number,
): RunningRow[] {
  const rows: RunningRow[] = [];

  for (const item of items) {
    const cost = item.cost ?? 0;
    if (cost <= 0) continue;
    const { shares, payerDisplay } = computeSharesForItem(item, people);
    rows.push({
      name: item.name.trim() || 'Unnamed Item',
      cost,
      payer: payerDisplay,
      shares: people.map((p, idx) => ({ name: p.name, amount: shares[idx] })),
    });
  }

  if (taxAmount > 0) {
    rows.push({
      name: 'Tax',
      cost: taxAmount,
      payer: 'Split by ratio',
      shares: people.map((p, idx) => ({
        name: p.name,
        amount: summary.taxShares[idx],
      })),
    });
  }

  if (tipAmount > 0) {
    rows.push({
      name: 'Tip',
      cost: tipAmount,
      payer: 'Split by ratio',
      shares: people.map((p, idx) => ({
        name: p.name,
        amount: summary.tipShares[idx],
      })),
    });
  }

  const grandTotal = round2(summary.totals.reduce((a, b) => a + b, 0));
  rows.push({
    name: 'TOTAL',
    cost: grandTotal,
    payer: '',
    shares: people.map((p, idx) => ({
      name: p.name,
      amount: summary.totals[idx],
    })),
    isTotal: true,
  });

  return rows;
}

File: app/streamlit_domain/running.py
from __future__ import annotations

from .math_utils import round2
from .models import Item, Person, RunningRow, RunningRowShare, SummaryResult
from .splits import compute_shares_for_item


def build_running_rows(
    people: list[Person],
    items: list[Item],
    summary: SummaryResult,
    tax_amount: float,
    tip_amount: float,
) -> list[RunningRow]:
    rows: list[RunningRow] = []

    for item in items:
        cost = item.cost or 0.0
        if cost <= 0:
            continue
        shares, payer_display = compute_shares_for_item(item, people)
        rows.append(
            RunningRow(
                name=item.name.strip() or "Unnamed Item",
                cost=cost,
                payer=payer_display,
                shares=[
                    RunningRowShare(name=person.name, amount=shares[idx])
                    for idx, person in enumerate(people)
                ],
            )
        )

    if tax_amount > 0:
        rows.append(
            RunningRow(
                name="Tax",
                cost=tax_amount,
                payer="Split by ratio",
                shares=[
                    RunningRowShare(name=person.name, amount=summary.tax_shares[idx])
                    for idx, person in enumerate(people)
                ],
            )
        )

    if tip_amount > 0:
        rows.append(
            RunningRow(
                name="Tip",
                cost=tip_amount,
                payer="Split by ratio",
                shares=[
                    RunningRowShare(name=person.name, amount=summary.tip_shares[idx])
                    for idx, person in enumerate(people)
                ],
            )
        )

    grand_total = round2(sum(summary.totals))
    rows.append(
        RunningRow(
            name="TOTAL",
            cost=grand_total,
            payer="",
            shares=[
                RunningRowShare(name=person.name, amount=summary.totals[idx])
                for idx, person in enumerate(people)
            ],
            is_total=True,
        )
    )

    return rows
